module-system/module-implementations/sov-evm/src/query.rs

```diff
@@ -25,42 +27,93 @@ impl<C: sov_modules_api::Context> Evm<C> {
         info!("evm module: eth_getBlockByNumber");
 
-        let header = reth_rpc_types::Header {
-            hash: Default::default(),
-            parent_hash: Default::default(),
-            uncles_hash: Default::default(),
-            miner: Default::default(),
-            state_root: Default::default(),
-            transactions_root: Default::default(),
-            receipts_root: Default::default(),
-            logs_bloom: Default::default(),
-            difficulty: Default::default(),
-            number: Default::default(),
-            gas_limit: Default::default(),
-            gas_used: Default::default(),
-            timestamp: Default::default(),
-            extra_data: Default::default(),
-            mix_hash: Default::default(),
-            nonce: Default::default(),
-            base_fee_per_gas: Some(reth_primitives::U256::from(100)),
-            withdrawals_root: Default::default(),
+        // "safe" and "finalized" are not implemented
+        let block = match block_number {
+            Some(ref block_number) if block_number == "earliest" => self
+                .blocks
+                .get(0, &mut working_set.accessory_state())
+                .expect("Genesis block must be set"),
+            Some(ref block_number) if block_number == "latest" => self
+                .head
+                .get(working_set)
+                .expect("Head block must be set")
+                .seal(),
+            Some(ref block_number) if block_number == "pending" => self
+                .pending_head
+                .get(&mut working_set.accessory_state())
+                .expect("Pending head block must be set")
+                .seal(),
+            Some(ref block_number) => {
+                let block_number = usize::from_str_radix(block_number, 16).unwrap();
+                self.blocks
+                    .get(block_number, &mut working_set.accessory_state())
+                    .expect("Block must be set")
+            }
+            None => self
+                .head
+                .get(working_set)
+                .expect("Head block must be set")
+                .seal(),
+        };
+
+        let (header, transactions) = {
+            let header = reth_rpc_types::Header::from_primitive_with_hash(block.header.clone());
+
+            let transaction_hashes = self
+                .transactions
+                .iter(&mut working_set.accessory_state())
+                .map(|tx| tx.signed_transaction.hash)
+                .collect::<Vec<_>>();
+
+            let tx_numbers = transaction_hashes
+                .iter()
+                .map(|hash| {
+                    (
+                        hash,
+                        self.transaction_hashes
+                            .get(hash, &mut working_set.accessory_state())
+                            .unwrap(),
+                    )
+                })
+                .collect::<HashMap<_, _>>();
```

Comment:
we shouldn't iterate over all transactions_hashes, that will load the entire database!

Reply:
That was my mistake, I confused block.transactions with db transactions. Fixed.